## Evaluation Exercises

### 2. Given the following confusion matrix, evaluate (by hand) the model's performance.

### - In the context of this problem, what is a false positive?
-- False Positive = predicted dog but is a cat

### - In the context of this problem, what is a false negative?
-- False Negative = predicted cat but is a dog

### - How would you describe this model?
-- accuracy = (TP + TN) / (TP + TN + FP + FN) = (46+34) / (46+34+13+7) = 80/100 = 80%

-- precision = TP / (TP + FP) = 46 / (46+13) = 46/59 = 77.97%

-- recall = TP / (TP + FN) = 46 / (46+7) = 46/53 = 86.79%

-- roughly 80% accuracy and precision means this model is good overall

-- 87% recall - high recall covers false negatives (predict dog but cat)

### 3. You are working as a datascientist working for Codeup Cody Creator (C3 for short), a rubber-duck manufacturing plant.

Unfortunately, some of the rubber ducks that are produced will have defects. Your team has built several models that try to predict those defects, and the data from their predictions can be found here.

In [2]:
import pandas as pd
import numpy as np

In [13]:
c3 = pd.read_csv('c3.csv')
c3.head()

actual     model1  model2     model3
0  No Defect  No Defect  Defect  No Defect
1  No Defect  No Defect  Defect     Defect
2  No Defect  No Defect  Defect  No Defect
3  No Defect     Defect  Defect     Defect
4  No Defect  No Defect  Defect  No Defect

In [14]:
c3.shape

(200, 4)

In [15]:
c3.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 200 entries, 0 to 199
Data columns (total 4 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   actual  200 non-null    object
 1   model1  200 non-null    object
 2   model2  200 non-null    object
 3   model3  200 non-null    object
dtypes: object(4)
memory usage: 6.4+ KB


### Use the predictions dataset and pandas to help answer the following questions:

- An internal team wants to investigate the cause of the manufacturing defects. They tell you that they want to identify as many of the ducks that have a defect as possible. Which evaluation metric would be appropriate here? Which model would be the best fit for this use case?

-- recall is the evaluation metric that is appropriate here to cover False Negatives

In [20]:
# positives are defective ducks
positive = 'Defect'

subset = c3[c3.actual == positive]
subset.head()

actual     model1     model2  model3
13  Defect  No Defect     Defect  Defect
30  Defect     Defect  No Defect  Defect
65  Defect     Defect     Defect  Defect
70  Defect     Defect     Defect  Defect
74  Defect  No Defect  No Defect  Defect

In [79]:
recall1 = (subset.actual == subset.model1).mean()
recall2 = (subset.actual == subset.model2).mean()
recall3 = (subset.actual == subset.model3).mean()

print(f'Recall rate for model1 is {recall1:.2%}')
print(f'Recall rate for model2 is {recall2:.2%}')
print(f'Recall rate for model3 is {recall3:.2%}')

Recall rate for model1 is 50.00%
Recall rate for model2 is 56.25%
Recall rate for model3 is 81.25%


-- model3's 81.25% recall rate is the best fit for this use case because it catches 81.25% of defective ducks

- Recently several stories in the local news have come out highlighting customers who received a rubber duck with a defect, and portraying C3 in a bad light. The PR team has decided to launch a program that gives customers with a defective duck a vacation to Hawaii. They need you to predict which ducks will have defects, but tell you they really don't want to accidentally give out a vacation package when the duck really doesn't have a defect. Which evaluation metric would be appropriate here? Which model would be the best fit for this use case?

-- precision is the best evaluation method to avoid giving expensive vacation when there really is no defect (False Positives)

In [25]:
# positves are defective ducks
positive = "Defect"

subset1 = c3[c3.model1 == positive]
subset2 = c3[c3.model2 == positive]
subset3 = c3[c3.model3 == positive]

precision1 = (subset1.model1 == subset1.actual).mean()
precision2 = (subset2.model2 == subset2.actual).mean()
precision3 = (subset3.model3 == subset3.actual).mean()

print(f'Precision rate for model1 is {precision1:.2%}')
print(f'Precision rate for model2 is {precision2:.2%}')
print(f'Precision rate for model3 is {precision3:.2%}')

Precision rate for model1 is 80.00%
Precision rate for model2 is 10.00%
Precision rate for model3 is 13.13%


-- model1 is the best fit for this use case because it will produce the least amount of False Positives

### 4. You are working as a data scientist for Gives You Paws ™, a subscription based service that shows you cute pictures of dogs or cats (or both for an additional fee).

### At Gives You Paws, anyone can upload pictures of their cats or dogs. The photos are then put through a two step process. First an automated algorithm tags pictures as either a cat or a dog (Phase I). Next, the photos that have been initially identified are put through another round of review, possibly with some human oversight, before being presented to the users (Phase II).

### Several models have already been developed with the data, and you can find their results here.

In [42]:
gyp = pd.read_csv('gives_you_paws.csv')
gyp.head()

actual model1 model2 model3 model4
0    cat    cat    dog    cat    dog
1    dog    dog    cat    cat    dog
2    dog    cat    cat    cat    dog
3    dog    dog    dog    cat    dog
4    cat    cat    cat    dog    dog

In [27]:
gyp.shape

(5000, 5)

### Given this dataset, use pandas to create a baseline model (i.e. a model that just predicts the most common class) and answer the following questions:

- a. In terms of accuracy, how do the various models compare to the baseline model? Are any of the models better than the baseline?

In [28]:
gyp.actual.value_counts()

dog    3254
cat    1746
Name: actual, dtype: int64

In [29]:
gyp['baseline'] = 'dog'

In [30]:
gyp.head()

actual model1 model2 model3 model4 baseline
0    cat    cat    dog    cat    dog      dog
1    dog    dog    cat    cat    dog      dog
2    dog    cat    cat    cat    dog      dog
3    dog    dog    dog    cat    dog      dog
4    cat    cat    cat    dog    dog      dog

In [37]:
# baseline accuracy
baseacc = (gyp.actual == gyp.baseline).mean()

# model1 accuracy
model1acc = (gyp.actual == gyp.model1).mean()

# model2 accuracy
model2acc = (gyp.actual == gyp.model2).mean()

# model3 accuracy
model3acc = (gyp.actual == gyp.model3).mean()

# model4 accuracy
model4acc = (gyp.actual == gyp.model4).mean()

print(f'Accuracy rate for the baseline is {baseacc:.2%}')
print(f'Accuracy rate for model1 is {model1acc:.2%}')
print(f'Accuracy rate for model2 is {model2acc:.2%}')
print(f'Accuracy rate for model3 is {model3acc:.2%}')
print(f'Accuracy rate for model4 is {model4acc:.2%}')

Accuracy rate for the baseline is 65.08%
Accuracy rate for model1 is 80.74%
Accuracy rate for model2 is 63.04%
Accuracy rate for model3 is 50.96%
Accuracy rate for model4 is 74.26%


-- models 1 and 4 had better accuracy than the baseline

#### b. Suppose you are working on a team that solely deals with dog pictures. Which of these models would you recomend for Phase I? For Phase II?

-- For Phase I, since this is the initial sorting of dogs and cats, accuracy would be the best metric to ensure highest rate of corrrectly sorting the 2 items.

-- For Phase II, recall would be the best metric

In [38]:
accmodel1 = (gyp.actual == gyp.model1).mean()
accmodel2 = (gyp.actual == gyp.model2).mean()
accmodel3 = (gyp.actual == gyp.model3).mean()
accmodel4 = (gyp.actual == gyp.model4).mean()

print(f'Accuracy rate for model1 is {accmodel1:.2%}')
print(f'Accuracy rate for model2 is {accmodel2:.2%}')
print(f'Accuracy rate for model3 is {accmodel3:.2%}')
print(f'Accuracy rate for model4 is {accmodel4:.2%}')

Accuracy rate for model1 is 80.74%
Accuracy rate for model2 is 63.04%
Accuracy rate for model3 is 50.96%
Accuracy rate for model4 is 74.26%


-- I would recommend model1 for Phase I because it has the highest accuracy rate.

In [46]:
# positives are dog
positive4 = 'dog'

subset4 = gyp[gyp.actual == positive4]

gyprecall1 = (subset4.actual == subset4.model1).mean()
gyprecall2 = (subset4.actual == subset4.model2).mean()
gyprecall3 = (subset4.actual == subset4.model3).mean()
gyprecall4 = (subset4.actual == subset4.model4).mean()

print(f'Recall rate for model1 is {gyprecall1:.2%}')
print(f'Recall rate for model2 is {gyprecall2:.2%}')
print(f'Recall rate for model3 is {gyprecall3:.2%}')
print(f'Recall rate for model4 is {gyprecall4:.2%}')

Recall rate for model1 is 80.33%
Recall rate for model2 is 49.08%
Recall rate for model3 is 50.86%
Recall rate for model4 is 95.57%


-- I would recommend model4 for Phase II because it has the highest recall rate.

#### c. Suppose you are working on a team that solely deals with cat pictures. Which of these models would you recomend for Phase I? For Phase II?

-- For Phase I, since this is (still) the initial sorting of dogs and cats, accuracy would be the best metric to ensure highest rate of corrrectly sorting the 2 items.

-- For Phase II, recall would be the best metric

In [47]:
accmodel1 = (gyp.actual == gyp.model1).mean()
accmodel2 = (gyp.actual == gyp.model2).mean()
accmodel3 = (gyp.actual == gyp.model3).mean()
accmodel4 = (gyp.actual == gyp.model4).mean()

print(f'Accuracy rate for model1 is {accmodel1:.2%}')
print(f'Accuracy rate for model2 is {accmodel2:.2%}')
print(f'Accuracy rate for model3 is {accmodel3:.2%}')
print(f'Accuracy rate for model4 is {accmodel4:.2%}')

Accuracy rate for model1 is 80.74%
Accuracy rate for model2 is 63.04%
Accuracy rate for model3 is 50.96%
Accuracy rate for model4 is 74.26%


-- I would (again) recommend model1 for Phase I because it has the highest accuracy rate.

In [48]:
# positives are dog
positive5 = 'cat'

subset5 = gyp[gyp.actual == positive5]

gyp5recall1 = (subset5.actual == subset5.model1).mean()
gyp5recall2 = (subset5.actual == subset5.model2).mean()
gyp5recall3 = (subset5.actual == subset5.model3).mean()
gyp5recall4 = (subset5.actual == subset5.model4).mean()

print(f'Recall rate for model1 is {gyp5recall1:.2%}')
print(f'Recall rate for model2 is {gyp5recall2:.2%}')
print(f'Recall rate for model3 is {gyp5recall3:.2%}')
print(f'Recall rate for model4 is {gyp5recall4:.2%}')

Recall rate for model1 is 81.50%
Recall rate for model2 is 89.06%
Recall rate for model3 is 51.15%
Recall rate for model4 is 34.54%


-- For Phase II with cats, model2 has the highest recall rate.

### 5. Follow the links below to read the documentation about each function, then apply those functions to the data from the previous problem.

- sklearn.metrics.accuracy_score

- sklearn.metrics.precision_score

- sklearn.metrics.recall_score

- sklearn.metrics.classification_report

In [49]:
from sklearn.metrics import accuracy_score
# computes the accuracy score, 0 to 1

In [57]:
acc5model1 = accuracy_score(gyp.actual, gyp.model1)
acc5model2 = accuracy_score(gyp.actual, gyp.model2)
acc5model3 = accuracy_score(gyp.actual, gyp.model3)
acc5model4 = accuracy_score(gyp.actual, gyp.model4)

print(f'Accuracy score for model1 is {acc5model1:.4}')
print(f'Accuracy score for model2 is {acc5model2:.4}')
print(f'Accuracy score for model3 is {acc5model3:.4}')
print(f'Accuracy score for model4 is {acc5model4:.4}')

Accuracy score for model1 is 0.8074
Accuracy score for model2 is 0.6304
Accuracy score for model3 is 0.5096
Accuracy score for model4 is 0.7426


In [60]:
from sklearn.metrics import precision_score
# computes precision score, 0 to 1

In [67]:
pre5model1 = precision_score(gyp.actual, gyp.model1, pos_label='dog')
pre5model2 = precision_score(gyp.actual, gyp.model2, pos_label='dog')
pre5model3 = precision_score(gyp.actual, gyp.model3, pos_label='dog')
pre5model4 = precision_score(gyp.actual, gyp.model4, pos_label='dog')

print(f'Precision score for model1 is {pre5model1:.4}')
print(f'Precision score for model2 is {pre5model2:.4}')
print(f'Precision score for model3 is {pre5model3:.4}')
print(f'Precision score for model4 is {pre5model4:.4}')

Precision score for model1 is 0.89
Precision score for model2 is 0.8932
Precision score for model3 is 0.6599
Precision score for model4 is 0.7312


In [65]:
from sklearn.metrics import recall_score
# computes recall score, 0 to 1

In [68]:
recall5model1 = recall_score(gyp.actual, gyp.model1, pos_label='dog')
recall5model2 = recall_score(gyp.actual, gyp.model2, pos_label='dog')
recall5model3 = recall_score(gyp.actual, gyp.model3, pos_label='dog')
recall5model4 = recall_score(gyp.actual, gyp.model4, pos_label='dog')

print(f'Recall score for model1 is {recall5model1:.4}')
print(f'Recall score for model2 is {recall5model2:.4}')
print(f'Recall score for model3 is {recall5model3:.4}')
print(f'Recall score for model4 is {recall5model4:.4}')

Recall score for model1 is 0.8033
Recall score for model2 is 0.4908
Recall score for model3 is 0.5086
Recall score for model4 is 0.9557


In [70]:
from sklearn.metrics import classification_report
# builds a report showing the main classification metrics

In [78]:
print('Classification report for model1')
print(classification_report(gyp.actual, gyp.model1))
print('------------------------------------------------------')
print('Classification report for model2')
print(classification_report(gyp.actual, gyp.model2))
print('------------------------------------------------------')
print('Classification report for model3')
print(classification_report(gyp.actual, gyp.model3))
print('------------------------------------------------------')
print('Classification report for model4')
print(classification_report(gyp.actual, gyp.model4))
print('------------------------------------------------------')

Classification report for model1
              precision    recall  f1-score   support

         cat       0.69      0.82      0.75      1746
         dog       0.89      0.80      0.84      3254

    accuracy                           0.81      5000
   macro avg       0.79      0.81      0.80      5000
weighted avg       0.82      0.81      0.81      5000

------------------------------------------------------
Classification report for model2
              precision    recall  f1-score   support

         cat       0.48      0.89      0.63      1746
         dog       0.89      0.49      0.63      3254

    accuracy                           0.63      5000
   macro avg       0.69      0.69      0.63      5000
weighted avg       0.75      0.63      0.63      5000

------------------------------------------------------
Classification report for model3
              precision    recall  f1-score   support

         cat       0.36      0.51      0.42      1746
         dog       0.66     